### PR TITLE
feat: add createFixedPeriodFromPeriodId method

### PR DESCRIPTION
**Pull request by Mohammer5**
Blocked by #12

### CODE REVIEW

**Review by HendrikThePendric — CHANGES_REQUESTED** [review]
In general, I wonder if the refactoring you did in https://github.com/dhis2/multi-calendar-dates/pull/12 wouldn't enable you to do this is a much smarter way. As you mention [here](https://github.com/dhis2/multi-calendar-dates/pull/12/files#r1106616516), you now have helper functions that create single period, not just helpers that create lists of periods....

If my understanding is correct, then this should allow you to get parse a period a bit like this:
1. Have an array of objects with a `regex` and `parse` property.
2. You loop through the array and if you've found a regex that matches the current period ID you've found the object you need.
3. Now you call the `parse` function on that object and pass it the period ID
4. In the `parse` function you extract the params needed by the `build**FixedPeriod` function from the period id and then call that builder function with the extracted params and return its return value.

If this works, I think it should be a lot less code. If it doesn't then we may have to reconsider the use of extracting those builder function in the first place.

Perhaps the above is naive, but in an ideal world it would work like this IMO.

**Comment by Mohammer5** [comment]
> If this works, I think it should be a lot less code. If it doesn't then we may have to reconsider the use of extracting those builder function in the first place.

It is possible, but not less code. The issue is that from the ID we need to figure out several things. One of those is the year of the period. If the period id is `2023NovemberS2`, the actual year of the period is 2024. I just gave it a try and it's more work than I expected.

> reconsider the use of extracting those builder function in the first place.

I think this is a fair assessment. Right now the only builder function that's used across all new features is `buildDailyFixedPeriod`, so we could keep that one extracted and remove all other builder utilities

**Review by HendrikThePendric — APPROVED** [review]
@Mohammer5 and I discussed yesterday that we can focus on getting stuff to work first and then clean up the implementation some more. 

My issue with the general approach, as I outline in https://github.com/dhis2/multi-calendar-dates/pull/13#pullrequestreview-1299828550 still remains the same: in https://github.com/dhis2/multi-calendar-dates/pull/12 a lot of effort (and LoC) was put into creating functions that could parse individual period codes, and if these were to work reliably we should be able to leverage these functions here. But it turns out this is (currently) not possible.

With our time constraints in mind, I think the following course of action would be best:
- I've approved now to unblock the PR
- @kabaros has left some comments which should be addressed by @Mohammer5 
- A JIRA issue for the follow-up work should be created by @Mohammer5. In the issue description there should be an explanation about the options we have:
    - Either we find a way to reliably parse individual periods directly from a period ID, and we leverage that here. If this were possible, it would definitely be the preferred option.
    - Or we reach the conclusion that parsing individual periods is not feasible, in which case we go back to the initial implementation, which basically means producing periods for an entire year and then identifying the correct one. If we go this way, we should probably revert some of the things we changed in #12. I am specifically thinking about this https://github.com/dhis2/multi-calendar-dates/pull/12#discussion_r1106616516

File: src/utils/get-first-day-number-of-month-by-calendar-type.ts [hunk 1]
```diff
@@ -0,0 +1,13 @@
+import { SupportedCalendar } from '../types'
+
+const getFirstDayNumberOfMonthByCalendarType = (
```
[the remark this answers is not in the document]
Comment (Mohammer5) [reply src/utils/get-first-day-number-of-month-by-calendar-type.ts]: removed

File: src/index.spec.ts [hunk 2]
```diff
@@ -9,6 +9,7 @@ describe('the interface of the library', () => {
                 'useDatePicker',
                 'useResolvedDirection',
                 'generateFixedPeriods',
+                'parseFixedPeriodId',
```
[the remark this answers is not in the document]
Comment (Mohammer5) [reply src/index.spec.ts:12]: I renamed it to `createFixedPeriodFromPeriodId`. I'm not a big fan of `get` as that implies that you have to get it from somewhere else (e.g. `getElementById` -> get it from dom, fetch method 'GET' -> get it from the server). My first thought was to use `generate` but that usually implies that some randomness is involved in the process

File: src/period-calculation/parse-fixed-period-id/get-weekly-fixed-period-type-for-period-id.ts [hunk 3]
```diff
@@ -0,0 +1,32 @@
+import { periodTypes } from '../period-types'
+import { PeriodType } from '../types'
+
+const getWeeklyFixedPeriodTypeForPeriodId = (periodId: string) => {
+    if (periodId.match(/^[0-9]{4}W[0-9]+$/)) {
```
[the remark this answers is not in the document]
Comment (Mohammer5) [reply src/period-calculation/parse-fixed-period-id/get-weekly-fixed-period-type-for-period-id.ts]: See the [`period-id`](https://github.com/dhis2/multi-calendar-dates/tree/parse-fixed-period-id/src/period-calculation/period-id) folder

File: src/period-calculation/parse-fixed-period-id/get-weekly-fixed-period-type-for-period-id.ts [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+    const weeklyOffsetPeriodMatch = periodId.match(
+        /^[0-9]{4}([A-Z][a-z]{2})W[0-9]+$/
+    )
+    if (weeklyOffsetPeriodMatch) {
+        const [, name] = weeklyOffsetPeriodMatch
+        const periodType = `WEEKLY${name.toUpperCase()}` as PeriodType
+
+        if (!periodTypes.includes(periodType)) {
+            throw new Error('@TODO')
```
[the remark this answers is not in the document]
Comment (Mohammer5) [reply src/period-calculation/parse-fixed-period-id/get-weekly-fixed-period-type-for-period-id.ts]: Now throws with a proper error message

**Pull request merged**